notebooks/ai-ml/SparkXGBoostCustomerChurn.ipynb

```diff
@@ -0,0 +1,838 @@
+   "outputs": [],
+   "source": [
+    "# Specify the model name and revision - downloads from Huggingface\n",
+    "model_name = \"distilbert-base-uncased-finetuned-sst-2-english\"\n",
+    "model_revision = \"af0f99b\" \n",
+    "\n",
+    "# Load pre-trained sentiment analysis model with specified model and revision\n",
+    "sentiment_pipeline = pipeline(\"sentiment-analysis\", model=model_name, revision=model_revision)\n",
+    "\n",
+    "# Define a UDF to perform sentiment analysis\n",
+    "@udf(returnType=StringType())\n",
+    "def analyze_sentiment(text):\n",
+    "    result = sentiment_pipeline(text)[0]\n",
+    "    return result['label']\n",
+    "\n",
+    "# Specify the path to the JSON file in GCS\n",
+    "file_path = f\"gs://{BUCKET}/unstructured-data/output/chat_transcripts.json\"\n",
+    "\n",
+    "# Read the JSON file into a Spark DataFrame\n",
+    "df_sentiment = spark.read.json(file_path)\n",
+    "\n",
+    "# Rename the \"support_transcript\" column to \"text\" (required for this particular model)\n",
+    "df_sentiment = df_sentiment.withColumnRenamed(\"support_transcript\", \"text\")\n",
+    "\n",
+    "# Remove records with null values for text\n",
+    "df_sentiment = df_sentiment.filter(df_sentiment[\"text\"].isNotNull() & (df_sentiment[\"text\"] != \"\"))\n",
+    "\n",
+    "# Analyze the \"support_transcript\" field for sentiment and retain \"customer_id\"\n",
+    "df_sentiment = df_sentiment.withColumn(\"sentiment\", analyze_sentiment(df_sentiment[\"text\"])).cache()\n",
+    "\n",
+    "# Display the results\n",
+    "df_sentiment.show()"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "0943683a-1aa7-4b6d-9a8a-846008b6076c",
+   "metadata": {
+    "tags": []
+   },
+   "source": [
+    "## Join data\n",
+    "\n",
+    "Combine the transformed transcript data to the customer churn data, then save this to Cloud Storage."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "id": "95e1b691-d216-43bd-9269-55c6a146c774",
+   "metadata": {
+    "tags": []
+   },
+   "outputs": [],
+   "source": [
+    "# Join the transformed and sentiment data, and save to GCS\n",
+    "\n",
+    "# Specify the output path in GCS\n",
+    "training_data_path = f\"gs://{BUCKET}/curated/data/training_data\"\n",
+    "\n",
+    "# Join the DataFrames on customer_id\n",
+    "df_combined = df_sentiment.join(df, on=\"CustomerID\", how=\"inner\") \n",
+    "\n",
+    "# # Write the DataFrame to GCS in Parquet format\n",
+    "df_combined.write.format(\"parquet\").mode(\"overwrite\").save(output_path)"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "a4867478-0467-4e29-8d49-ab0b4e41a197",
+   "metadata": {},
+   "source": [
+    "# Model Training and Evaluation\n",
+    "\n",
+    "Next, you train an XGBoost model using the newly processed data. \n",
+    "\n",
+    "XGBoost does not support Spark's dynamic allocation feature. To use XGBoost, disable this first by restarting the Spark Session."
```

Comment:
Perhaps explain in a sentence what Spark's dynamic allocation feature is (or link to it), and _why_ this is a limitation or what it means for this type of job.